### ISSUE TITLE
Invalid url for method `GoogleApi.Storage.V1.Api.Objects.storage_objects_insert`

### ISSUE THREAD

**Issue by charlesdurand** [googleapis/elixir-google-api#972]
I believe it should be `/upload/storage/v1/b/{bucket}/o` instead of `/storage/v1/b/{bucket}/o`

https://github.com/googleapis/elixir-google-api/blob/dac4877b5da2694eca6a0b07b3bd0e179e5f3b70/clients/storage/lib/google_api/storage/v1/api/objects.ex#L432

**Comment by chingor13** [comment]
You actually want to use storage_objects_insert_simple which has the correct URL and can specify the uploadType parameter.

**Comment by EskiMag** [comment]
So the `storage_objects_insert` function has incorrect URL and is not working? Or is it somehow intended for different use cases? I am a bit confused.
I am still not sure how to address the #1197 
As a workaround, I am using a modified fork of this repo as a local dependency with the `upload` URL prefix, which is working OK for me.
There was no update after the GCP Next conference as @dazuma said in my issue, so I am not sure which approach to choose.
Thanks a lot for directions.

**Comment by vjebelev** [comment]
> You actually want to use `storage_objects_insert_simple` which has the correct URL and can specify the `uploadType` parameter.

storage_objects_insert_simple() is designed to work with file names only while storage_objects_insert() allows you to upload a string from memory, which is valuable. Please fix the url with the next regen.

**Comment by kolorahl** [comment]
@chingor13 I believe the problem is a disconnect between the REST specification as returned by the discovery API and the underlying API implementation. 

## TL;DR

The storage API needs to accept `POST /b/{bucket}/o` as a valid path and, ideally, use `uploadType` to determine which protocol to use (`simple`, `multipart`, and `resumable`) rather than having multiple paths that map to different upload types (`simple` and `multipart` use the same path but `resumable` uses a different one). But because it will likely take a long time for the team managing that API to make the necessary fixes, I have two possible solutions for the interim:

1. It's not my ideal solution, but it would be rather simple to look for an id as a key in a configuration map (e.g. `storage.objects.insert`) and merge those overrides on top of the discovered metadata. This would require us to maintain a configuration map somewhere but it allows us to alter the metadata used to generate code without messing up the actual discovery or code generation components, i.e. if no overrides exist then behavior would remain exactly as it is now, and if overrides _do_ exist then the behavior still remains the same except the input is transformed before generation starts.

2. A perhaps more dynamic solution would be to allow practically any value to be overridden from an optional keyword list or through application configs. For example, the `opts` keyword list could allow for things like `base_path`, `path_prefix`, and/or `library_version` that would override the default values set by the generator.

Ultimately it is a matter of whether we want the output (i.e. client code) to remain the same (option 1) or to change to provide more dynamic behavior (option 2).

## The Whole Story

I understand these clients are being generated from a Swagger API specification, and looking at the specification I see the following:

```json
"insert": {
  "id": "storage.objects.insert",
  "path": "b/{bucket}/o",
  "httpMethod": "POST",
  "..."
}
```

The [official API reference](https://cloud.google.com/storage/docs/json_api/v1/) says that path should theoretically be correct:

> POST
> https://www.googleapis.com/upload/storage/v1/b/bucket/o
> and
> POST  /b/bucket/o

But the error message returned by the server says the latter version (`/b/bucket/o`) is not valid:

> "message": "Upload requests must include an uploadType URL parameter and a URL path beginning with /upload/"

Also, `uploadType` does not appear to be a required query parameter despite the specification and error message both mentioning that it is. The [endpoint generation code](https://github.com/googleapis/elixir-google-api/blob/master/lib/google_apis/generator/elixir_generator/endpoint.ex) will manually add `uploadType` as a required query parameter for the `simple` and `resumable` protocols, but the basic endpoint does not have `uploadType` set as a required parameter in the specification; only `bucket` is a required parameter. In fact, the specification doesn't have `uploadType` in the parameter list at all, not even as an optional parameter, so you can't set it in `storage_objects_insert/4` even if you wanted to. Despite this, the call still succeeds as long as the path is correct.

Ultimately, Google is not respecting their own API specification (they fail on the resource path listed in their own specification) and have misleading error messages (they communicate rules about valid requests that aren't completely respected by the server). Either the specification needs to be corrected, the implementation needs to be corrected, or both need to be adjusted to bring it to a happy medium. Barring that, the generator might need hard-coded logic to create a workaround for this and other instances where the specification and implementation do not agree.

In case there's any disagreement on whether the specification and implementation do not agree, or whether we should be "hard-coding" solutions to this problem, I would like to redirect any interested parties to the [base service object for Ruby](https://github.com/googleapis/google-api-ruby-client/blob/711dfb83b33c03535076917726956584d5c8bf9a/lib/google/apis/core/base_service.rb). If you look at lines 105 and 131, you can see that the non-generated `BaseService` class definition has an "override" attribute called `upload_path` that is **specifically aimed at modifying the path for media uploads**. The value is hard-coded to use the `/upload/` prefix. So clearly this problem exists in other versions of the client libraries but are being overridden manually to adjust for the path discrepancy.

**Comment by dazuma** [comment]
I did some digging here, and the issue is both a bit simpler and more complicated. Simpler in that we've done away with Swagger and are now generating directly from the source Discovery document, which does in fact have the upload URLs available (so we should be able to fix this without requiring user override of the path.) But more complicated in that we can't just change the URL for `storage_objects_insert` since it's being generated for a different case (metadata-only upload—and just to make it more confusing, Storage doesn't actually support this operation, unlike other similar upload APIs such as Drive and YouTube.) To actually upload data, the intent was to generate an auxiliary function `storage_objects_insert_simple`, as @chingor13 points out. Unfortunately, as you point out, this function as written requires you to upload from a file.

So what we're going to do is generate a second auxiliary function `storage_objects_insert_iodata`. This will be the same as `storage_objects_insert_simple` except it will take an iodata (i.e. either a raw string/binary, or an iolist) instead of a filename. A PR is open for this change; we'll try to get the storage client regenerated with the addition later this week.

**Comment by dazuma** [comment]
The `google_api_storage` package should be released by Monday. Use the `storage_objects_insert_iodata` function.

**Comment by ESearcy** [comment]
> I did some digging here, and the issue is both a bit simpler and more complicated. Simpler in that we've done away with Swagger and are now generating directly from the source Discovery document, which does in fact have the upload URLs available (so we should be able to fix this without requiring user override of the path.) But more complicated in that we can't just change the URL for `storage_objects_insert` since it's being generated for a different case (metadata-only upload—and just to make it more confusing, Storage doesn't actually support this operation, unlike other similar upload APIs such as Drive and YouTube.) To actually upload data, the intent was to generate an auxiliary function `storage_objects_insert_simple`, as @chingor13 points out. Unfortunately, as you point out, this function as written requires you to upload from a file.
> 
> So what we're going to do is generate a second auxiliary function `storage_objects_insert_iodata`. This will be the same as `storage_objects_insert_simple` except it will take an iodata (i.e. either a raw string/binary, or an iolist) instead of a filename. A PR is open for this change; we'll try to get the storage client regenerated with the addition later this week.

This is still not working.,
Whats the deal?? some of us actually want to use the libraries you guys spend so much time working on.
this path issue still hasn't been figured out?

like... why is this issue Closed? The issue still exists!!

**Comment by dazuma** [comment]
@ESearcy What issue are you referring to? "Not working" is not a helpful bug report.

If you're talking about #3407, please comment there, or on the PR #3638 that is open to address it.

If you're not talking about #3407, please open a new issue and explain what your problem is.

Thanks.

**Comment by zachdaniel** [comment]
I think what the previous person is pointing out is that the function `storage_objects_insert` is still defined and is still using the incorrect url. So even though there is a different function to use, it would stand to reason that the function that doesn't work should be fixed or removed.
